src/Autofac.Integration.WebApi.Owin/DependencyScopeHandler.cs

```diff
@@ -42,13 +42,19 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
         var dependencyScope = new AutofacWebApiDependencyScope(lifetimeScope);
         request.Properties[HttpPropertyKeys.DependencyScope] = dependencyScope;
 
-        try
-        {
-            return base.SendAsync(request, cancellationToken);
-        }
-        finally
-        {
-            request.Properties.Remove(HttpPropertyKeys.DependencyScope);
-        }
+        // Using .ContinueWith instead of try/finally because you can't
+        // have async/await in [SecurityCritical] or [SecuritySafeCritical]
+        // marked code.
+        return base
+            .SendAsync(request, cancellationToken)
+            .ContinueWith(
+                task =>
+                {
+                    request.Properties.Remove(HttpPropertyKeys.DependencyScope);
+                    return task.Result;
+                },
+                CancellationToken.None,
+                TaskContinuationOptions.ExecuteSynchronously,
+                TaskScheduler.FromCurrentSynchronizationContext());
```

Comment:
Does `TaskScheduler.FromCurrentSynchronizationContext()` make sense when we use `TaskContinuationOptions.ExecuteSynchronously`?

Reply:
That's a good question. Reading the docs on the [analyzer](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2008) I got into [this blog article](https://devblogs.microsoft.com/pfxteam/new-taskcreationoptions-and-taskcontinuationoptions-in-net-4-5/) that seemed to recommend this as the way, but I suppose we could use `TaskScheduler.Current` to continue on the same thread, yeah? Again, my `Task` is a little rusty. ;)

Reply:
I would probably just use the overload that needs only `TaskContinuationOptions` (it will still pass `TaskScheduler.Current`, but that will be out of our hands). I cannot check whether this will still trigger the analyzer, but if it will, I think we can safely suppress the warning saying "we don't care, really don't care where it runs".

Reply:
Ended up having to _add updated analyzers_ to see the analyzer fire... then found that it does get caught, so suppressed it. Which seems sort of weird but I'd rather know we thought about it and explicitly chose not to do anything rather than it seeming like we didn't think about it at all.

Reply:
Feel like a terrible person saying it, but isn't there an overload of `ContinueWith` that takes a CancellationToken? I think you need to pass the existing `cancellationToken` into the continuation to ensure the returned Task can get IsCanceled set correctly if the token gets cancelled.

Arguably it doesn't matter if the continuing action is short and synchronous, but I can't remember whether the original task's cancellation state is passed through to the returned Task.

Manual ContinueWith sucks.

Reply:
It seems like, from a `ContinueWith` overload perspective, you can either specify _almost nothing_ or _almost everything_ but not really both. What that means is:

- I added the original `cancellationToken` passing through, _but_
- I also now have to specify `TaskScheduler.Current` and can't leave it to defaulted, which means
- I don't have to suppress the analyzer anymore

If y'all could just move off old Web API and OWIN this wouldn't be an issue. 😆